### PR TITLE
Check the invalid dimensions for matrix symbol creation

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

Previously, matrix did accept any dimensions, but I'm attempting to add some sanity check.
I've added a strict check for numerics, and a loose check for symbolic dimensions (like rejecting an obvious assumptions like `integer=False`, but doesn't strictly have to be `integer=True, nonnegative=True` or such, regarding the possibility that real symbol can substituted with integer)

I've also made a more strict check about `Lambda` because it is making the `Basic` object to contain non-sympy objects.

#### Release Notes

- matrices
  - `ZeroMatrix`, `OneMatrix`, `Identity`, `DFT`, `FuncMatrix` will give `ValueError` when invalid  dimensions (non-integers or negative or non-real) are given.
  - `FuncMatrix` will only accept SymPy `Lambda` or a string form of python `lambda`

### CODE REVIEW

**Comment by oscargus** [comment]
I'm not sure about the Lambda constraint. `sympify` with provide a `Lambda` for `sympify("lambda x: x + 1")`and while I'm not a big fan of entering strings, I think that at least one at least can do a proper `sympify` here. 

Also, I was under the impression that `_sympify` primarily should be used for library code, not for end-user code, but I realize that there is quite a bit of `_sympify` around.

File: sympy/matrices/expressions/matexpr.py [hunk 1]
```diff
@@ -227,6 +227,15 @@ def _visit_eval_derivative_array(self, x):
     def _accept_eval_derivative(self, s):
         return s._visit_eval_derivative_array(self)
 
+    @classmethod
+    def _check_dim(cls, dim):
```
Comment (smichr) [inline sympy/matrices/expressions/matexpr.py]: I wonder if `check_assumptions` could be used here (and elsewhere more frequently):
```python
ok = check_assumptions(dim, integer=True, nonnegative=True)
if ok is False:  raise ValueError
```

Comment (smichr) [reply sympy/matrices/expressions/matexpr.py]: (None will be returned as long as no assumption was False when it should have been True.)

**Comment by sylee957** [comment]
I see there is a parser to handle lambda string as a whole
If it is possible, I'd suppose that it can figure out any side effects.

**Comment by sylee957** [comment]
I'm not sure what is happening with the test failings with `tree` printer, but it seems like the assumptions are built or sorted in some randomized manner.

**Comment by oscarbenjamin** [comment]
Looks good to me

**Comment by sylee957** [comment]
I'm just going to make some more pushes to make the coverage 100%. I'll merge this after finished.

**Comment by sylee957** [comment]
Also, I have disallowed other symbolic functions than `Lambda` here
because I couldn't find any way to check if a symbolic function can accept 2 arguments.

**Comment by oscarbenjamin** [comment]
There is `nargs`:
```julia
In [2]: sin.nargs                                                                                                                 
Out[2]: {1}

In [3]: log.nargs                                                                                                                 
Out[3]: {1, 2}

In [4]: atan2.nargs                                                                                                               
Out[4]: {2}
```

**Comment by oscarbenjamin** [comment]
Looks good

**Pull request merged**